ports/PWM: Reduce inconsistencies between the ports.

robert-hh opened the pull request:
This is the first set of commits, consisting of:

- mimxrt: Start PWM only if freq and duty are set. And straighten the duty handling.
- samd: Add the init() method to PWM. This is merely restructuring  existing code, at some costs for flash size and RAM usage. init() without argument restarts a PWM channel which had been stopped with deinit().
- rp2: Enable keyword arguments with the PWM constructor and add the init() method. init() restarts a PWM channel which had been stopped with deinit(). Added code to stop PWM on soft reset.
- Adapt the quickref examples to set freq and duty_u16 using keyword arguments in the constructor.
- ESP8266: Add the methods duty_u16() and duty_ns().
- rp2: Edd a invert=True|False keyword argument to init() and the constructor to invert channel. In order to use not more memory, some flags of the object are put into a bit field.

So Rp2, ESP8266, MIMXRT and SAMD support now keyword arguments with the constructor and an init() method with optional argumnents. All four do no start PWM until freq and duty are set. The behavior at deinit() is different. The state of the pins during set-up and after deinit() varies. For ESP32 there is a PR enforcing freq and duty settings with the constructor or init(). 

The STM32 and Renesas-RA ports do not have a machine.PWM class, but it could be added. The nrf port's PWM modules behaves completely different and requires a full rework to get compatible.

robert-hh committed:
mimxrt/machine_pwm: Start PWM only if freq and duty are set.

And also fix/improve the following:
- Simplify the duty handling a little bit.
- Allow duty_u16(65536), which sets the output high.
- Rename machine_pwm_start() to mp_machine_pwm_start(), in preparation for
  a possible start/stop method pair.

robert-hh committed:
samd/machine_pwm: Add init() method to PWM and simplify the PWM code.

The PWM.init() method has been added.  Calling init() without arguments
restarts a PWM channel stopped with deinit().  Otherwise single parameters
except for "device=n" can be changed again.  The device can only be
specified once, either in the constructor or the first init() call.

Also simplify get_pwm_config() and get_adc_config(), and shrink the PWM
object.

robert-hh committed:
rp2/machine_pwm: Enable keyword args in constructor and add init method.

This adds support for freq/duty_u16/duty_ns keyword arguments in the PWM
constructor, and adds the PWM.init() method.  Using init() without
arguments enables a previously deinit-ed PWM again.

Further changes in this commit:
- Do not start PWM output if only duty was set.
- Stop all PWM slices on soft-reset.
- Fix a bug when changing the freq on a channel pair with duty_ns set.

robert-hh committed:
docs/samd: Make use of pin names more consistent in examples.

This keeps up with the changed Pin naming scheme.

robert-hh committed:
docs: Update the PWM examples based on recent API improvements.

This adds the freq and duty_u16 keyword settings to the constructor, and
sometimes other details in the PWM section.

For mimxrt a clarification regarding the PWM invert argument was added, and
for rp2 a few words were spent on PWM output pairs of a channel/slice.

robert-hh committed:
esp8266/machine_pwm: Implement duty_u16() and duty_ns() for consistency.

Also adds these keyword arguments to the constructor and init method.

robert-hh committed:
rp2/machine_pwm: Add support for inverting a PWM channel output.

Using the invert=True|False keyword option with the constructor or init().

robert-hh commented:
I could not stand to leave the nrf port aside and so I started to rewrite the machine.PWM implementation. hard PWM looks fine so far, soft PWM has to be adapted, and more testing is due. But I'll make  a separate PR for it.

robert-hh commented:
This PR is good for now as a first step of reducing the difference between ports. I have prepared another PR for adding the methods start() and stop() to PWM, but the semantics must be still discussed. 
The ESP32 change is still going on in a separate PR.

DavidEGrayson commented:
It looks like you're revamping the way PWM works, so I'd just like to request a feature for the RP2 port: we should be able to initialize the PWM module and set its properties and then later connect it to a pin.  This comes up when using inverted PWM to control an LED that is connected between an I/O pin and VDD: you want to set up the PWM module and make sure its output is driving high *before* connecting that to the pin.  Currently, we can't do that, so the pin would have to drive low briefly when setting up the PWM, making the LED flash.

robert-hh commented:
Thank you for the hint. The actual state of that PR is having the same API and basic behavior through all major ports. One open topic which has to be defined is the state of pins before PWM starts and after it has stopped, e.g. after calling PWM.deinit(). Moving the initilisation of the pin to the latest possible moment may be an option.

robert-hh committed:
rp2/machine_pwm: Add duty_x() checks and return 0 if PWM is not started.

Changes in this commit:
- Limit duty_u16() to 65535 and duty_ns() to the period duration.
- Return 0 for pwm.freq() if the frequency has not been set yet.
- Return 0 for pwm.duty_us16() and duty_ns() unless both frequency and
  duty cycle have been set.
- Initialize the pin to PWM at the very end of the constructor, to avoid
  possible glitches on the pin when setting up the PWM.

pdg137 commented on ports/rp2/machine_pwm.c:
```diff
@@ -207,7 +299,7 @@ STATIC void mp_machine_pwm_duty_set_ns(machine_pwm_obj_t *self, mp_int_t duty_ns
         mp_raise_ValueError(MP_ERROR_TEXT("duty larger than period"));
```
Not part of your changes, but as long as you are going over things, this error message should change.  It's not checking it against the period (given by TOP), just against the maximum possible value of 65535.

Also the maximum value depends unpredictably on internal details of the PWM implementation.  To make it more predictable and consistent across slight implementation changes, I think we should just cap the value instead of raising an error here.

```python
MicroPython v1.20.0-31-g3e565deee on 2023-04-28; Raspberry Pi Pico with RP2040

Type "help()" for more information.

>>> from machine import PWM, Pin
p=PWM(Pin(25))
p.freq(1000000)
>>> p.duty_ns(1000) # 100% duty
>>> p.duty_ns(500000) # 50,000% duty
>>> p.duty_ns(600000) # 60,000% duty
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: duty larger than period
>>> 
```

dpgeorge replied:
This error has been removed.

pdg137 commented on ports/rp2/machine_pwm.c:
```diff
@@ -165,14 +244,24 @@ STATIC void mp_machine_pwm_freq_set(machine_pwm_obj_t *self, mp_int_t freq) {
+    }
 }
 
 STATIC mp_obj_t mp_machine_pwm_duty_get_u16(machine_pwm_obj_t *self) {
+    if (self->duty_type == DUTY_NOT_SET || slice_freq_set[self->slice] == false) {
+        mp_raise_ValueError(MP_ERROR_TEXT("PWM not configured"));
+    }
```
I don't like this situation where then PWM module has been initialized and is doing *something*, but I'm not allowed to ask what it's doing:

```python
from machine import PWM, Pin
p=PWM(Pin(25)) # Pin 25 goes low, LED turns on
p.duty_u16(65535) # duty cycle is still 0%
p.duty_u16() # error
p.freq() # error
```

Just as an example of how this could be a disaster in real applications, imagine a display update that happens somehow asynchronously and checks `duty_u16()` as part of the update cycle for some reason. Maybe it's a servo controller that outputs the current servo settings periodically to a serial port. Usually it's fine but 1 out of 1000 times the display update happens before duty/freq are both set. Now we'll have an error maybe causing a crash in the field where it can't be debugged.

Note that this hypothetical application might already be written. We don't have a way to notify the author of the new requirements, and this update will silently cause their code (which worked fine before) to become unreliable.

It seems like it would be much better to allow checking PWM state without errors.  I would much prefer one of these three options:

* start at a reasonable default frequency (the current behavior for rp2)
* return zero duty cycle and frequency when PWM is off
* require setting an initial frequency and duty cycle in the constructor

As a comparison, Arduino's [analogWrite()](https://www.arduino.cc/reference/en/language/functions/analog-io/analogwrite/) function uses a preconfigured frequency. I had expected MicroPython to work the same way when I got the surprising result I mentioned in #11336.

dpgeorge replied:
The duty and freq methods no longer raise exceptions when queried, but instead return 0 if not started yet.

robert-hh commented:
Raising an error may be a little bit harsh. Instead, 0 could be returned. About the options, requesting freq and duty rate to be set with the constructor results is the most consistent option, but is a breaking change.

> start at a reasonable default frequency (the current behavior for rp2)

The actual behavior of the RP2 port is to start at a frequency of ~1907.35 Hz, after the duty rate has been set, but not the ferquency. I would not call that a reasonable default frequency. That's just the result of top being 0xffff and div=1. And that behavior is considered as a fault.

As of the other comment: I prepared a change to limit duty_ns(ns) to the top value,  and duty_u16 to 65535.

I still to not understand your condern about the duty rate of 65535.

pdg137 commented:
> requesting freq and duty rate to be set with the constructor results is the most consistent option, but is a breaking change.

Requiring frequency to be set the way you are doing it now is also a breaking change.  For example it breaks the first part of the first example [in the PWM docs](https://docs.micropython.org/en/latest/library/machine.PWM.html):

```python
pwm = PWM(pin)          # create a PWM object on a pin
pwm.duty_u16(32768)     # set duty to 50%
```

Also, did I understand correctly that you're planning to require freq/duty in the constructor for the ESP32 port? If so, shouldn't rp2 also require those arguments for consistency? At least it's not likely to be a silent/intermittent error.

> I still to not understand your concern about the duty rate of 65535.

Is this in reference to my other PR #11336? It seems like it's going to get addressed one way or another here so I'm not concerned anymore.

robert-hh commented:
> Requiring frequency to be set the way you are doing it now is also a breaking change.

I consider the actual behavior of the RP2 port a bug, in starting at an arbitrary frequency (1907.35 Hz) if only the duty rate is set. PWM must only start if both freq and duty rate are set.

> Also, did I understand correctly that you're planning to require freq/duty in the constructor for the ESP32 port?

No. I do not plan that. That was just an option which you suggested. It could be required, and one aim of this PR is to allow for all ports setting both frequency and duty rate with the constructor. The actual ESP32 port (and the PR for the update) behaves different, in that it starts PWM with 5KHz/50% duty if the constructor is called with just the Pin argument. That is different to all other ports and should be changed.

DavidEGrayson commented on ports/rp2/machine_pwm.c:
```diff
@@ -87,16 +139,40 @@ STATIC mp_obj_t mp_machine_pwm_make_new(const mp_obj_type_t *type, size_t n_args
     self->duty_type = DUTY_NOT_SET;
 
     // Select PWM function for given GPIO.
     gpio_set_function(gpio, GPIO_FUNC_PWM);
```
This line should be removed since it causes glitches like I was describing, and you already copied it to the end of the function.

dpgeorge replied:
This was fixed.

dpgeorge committed:
extmod/machine_pwm: Remove PWM_INIT and PWM_DUTY_U16_NS config options.

All ports that enable MICROPY_PY_MACHINE_PWM now enable these two
sub-options, so remove these sub-options altogether to force consistency in
new ports that implement machine.PWM.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
This is looking good, a good improvement to over consistency between ports PWM behavoiur.  Thank you very much @robert-hh !

I have cleaned up the commits a bit (splitting out the non-PWM docs changes to a separate commit), and added a commit that removes `MICROPY_PY_MACHINE_PWM_INIT` and `MICROPY_PY_MACHINE_PWM_DUTY_U16_NS` because they are now always enabled by all ports.

I also removed the commented-out code in `rp2/machine_pwm.c`:
```diff
--- a/ports/rp2/machine_pwm.c
+++ b/ports/rp2/machine_pwm.c
@@ -158,15 +158,7 @@ void machine_pwm_deinit_all(void) {
     for (int i = 0; i < 8; i++) {
         slice_freq_set[i] = false;
         pwm_set_enabled(machine_pwm_obj[i].slice, false);
-        // pwm_set_chan_level(i, 0, 0);
-        // pwm_set_chan_level(i, 1, 0);
     }
-    // Clean out the table
-    // for (int i = 0; i < 16; i++) {
-    //     machine_pwm_obj[i].invert = 0;
-    //     machine_pwm_obj[i].duty = 0;
-    //     machine_pwm_obj[i].duty_type = DUTY_NOT_SET;
-    // }
 }
 
 STATIC void mp_machine_pwm_deinit(machine_pwm_obj_t *self) {
```

robert-hh commented:
Thank you for merging. It's now only the ESP32 port which behaves a little bit different in starting with a default frequency/duty when these are not set at instantiation. The author insists on the way it behaves.

IhorNehrutsa commented:
@robert-hh @dpgeorge ESP32 port will be changed in #10854
